doc/architectural/goto-program-transformations.md

```diff
@@ -0,0 +1,311 @@
+
+This is a pass that can add many checks and instrumentation largely related to
+the definition of the C language. Note that this transformation pass is
+mandatory in the current implementation, however it may have no effect depending
+on the command line options specified.
+
+The implementation of this pass is called via \ref
+goto_check_c(const optionst &, goto_modelt &, message_handlert &)
+
+<em>Predecessor pass is \ref unions-transform.</em>
+
+
+\subsection floats-transform Adjust Float Expressions
+
+This is a transform from general mathematical operations over floating point
+types into floating point specific operator variations which include a rounding
+mode.
+
+The implementation of this pass is called via \ref
+adjust_float_expressions(goto_modelt &) . Documentation of this pass can be
+found in \ref adjust_float_expressions.h
+
+<em>Predecessor pass is \ref check-c-transform or the optional \ref
+assertions-transform if it is being used.</em>
+
+
+\subsection update-transform Goto Functions Update
+
+This transformation updates in memory data goto program fields which may
+have been invalidated by previous passes. Note that the following are performed
+by this pass:
+- Incoming edges
+- Target numbers
+- location numbers
+- loop numbers
+
+The implementation of this pass is called via \ref goto_functionst::update()
+
+<em>Predecessor pass is \ref floats-transform or the optional \ref
+string-abstraction-transform if it is being used.</em>
+
+
+\subsection failed-symbols-transform Add Failed Symbols
+
+This pass adds failed symbols to the symbol table. See
+`src/pointer-analysis/add_failed_symbols.h` for details. The implementation of
+this pass is called via \ref add_failed_symbols(symbol_table_baset &) . The
+purpose of failed symbols is explained in the documentation of the function \ref
+goto_symext::dereference(exprt &, statet &, bool)
+
+<em>Predecessor pass is \ref update-transform or the optional \ref
+nondet-transform if it is being used.</em>
+
+
+\subsection remove-skip-transform Remove Skip Instructions
+
+This transformation removes skip instructions. Note that this transformation is
+called in many places and may be called as part of other transformations. This
+instance here is part of the mandatory transformation to reach core goto format.
+If there is a use case where it is desirable to preserve a "no operation"
+instruction, a `LOCATION` type instruction may be used in place of a `SKIP`
+instruction. The `LOCATION` instruction has the same semantics as the `SKIP`
+instruction, but is not removed by the remove skip instructions pass.
+
+The implementation of this pass is called via \ref remove_skip(goto_modelt &)
+
+<em>Predecessor pass is \ref failed-symbols-transform or the optional \ref
+unused-functions-transform if it is being used.</em>
+
+
+
+\subsection properties-transform Label Properties
+
+This transformation adds information to assert instructions. This includes
```

Comment:
The need for this transform is troubling. How are we letting in assertions that aren't annotated with source information? I think the right solution is to start forcing code to have locations when it is added.

Reply:
So can we just reword this to emphasize what is expected? Maybe:

> This transformation adds unique IDs to the location information for a property. It will also infer and populate file and line information from the surrounding context if this is missing.

Reply:
Ok. I think I am going to remove any reference to location addition from this section. The pass doesn't infer any file/line information. It just adds the blank data structure, so that the property IDs can be added. The file, line, function, column etc. members of the location structure will remain empty. The location structure addition is just an implementation detail of adding the property identifier in this case.